eval/bundled/epm.elv.go

```diff
@@ -1,111 +1,279 @@
+        -error "Package "$pkg" is not installed."
         return
+      }
+      -info "Updating "$pkg
+      rsync -av ($-method-handler[rsync][src] $pkg $dom-cfg) $dest
     }
-    -info 'Upgrading package '$pkg
-    git -C $dest pull
+  ]
+]
+
+# Return the filename of the domain config file for the given domain
+# (regardless of whether it exists)
+fn -domain-config-file [dom]{
+  put $-lib-dir/$dom/epm-domain.cfg
 }
 
-fn upgrade [@pkgs]{
-    if (eq $pkgs []) {
-        pkgs = [(installed)]
-        -info 'Upgrading all installed packages'
-    }
-    for pkg $pkgs {
-        -upgrade-one $pkg
+# Return the filename of the metadata file for the given package
+# (regardless of whether it exists)
+fn -package-metadata-file [pkg]{
+  put (dest $pkg)/metadata.json
+}
+
+# Read the domain config file for a given domain. If the file does not
+# exist but we have a built-in definition, then we create the file
+# with the default.
+fn -read-domain-config [dom]{
+  cfgfile = (-domain-config-file $dom)
+  # Only read config if it hasn't been loaded already
+  if (not (has-key $-domain-config $dom)) {
+    if ?(test -f $cfgfile) {
+      # If the config file exists, read it...
+      -domain-config[$dom] = (cat $cfgfile | from-json)
+      -debug "Read domain config for "$dom": "(to-string $-domain-config[$dom])
+    } else {
+      # ...otherwise check if we have a default config for the domain, and save it
+      if (has-key $-default-domain-config $dom) {
+        -domain-config[$dom] = $-default-domain-config[$dom]
+        -debug "No existing config for "$dom", using the default: "(to-string $-domain-config[$dom])
+        mkdir -p (dirname $cfgfile)
+        put $-domain-config[$dom] | to-json > $cfgfile
+      } else {
+        fail "No existing config for "$dom" and no default available. Please create config file "(tilde-abbr $cfgfile)" by hand"
+      }
     }
+  }
 }
 
-fn -uninstall-one [pkg]{
-    installed-pkgs = [(installed)]
-    if (not (has-value $installed-pkgs $pkg)) {
-        -error 'Package '$pkg' is not registered as installed.'
-        return
+# Return the method by which a package is installed
+fn -package-method [pkg]{
+  dom = (-package-domain $pkg)
+  -read-domain-config $dom
+  if (has-key $-domain-config $dom) {
+    cfg = $-domain-config[$dom]
+    put $cfg[method]
+  } else {
+    put $false
+  }
+}
+
+# Invoke package operations defined in $-method-handler above
+fn -package-op [pkg what]{
+  dom = (-package-domain $pkg)
+  -read-domain-config $dom
+  if (has-key $-domain-config $dom) {
+    cfg = $-domain-config[$dom]
+    method = $cfg[method]
+    if (has-key $-method-handler $method) {
+      if (has-key $-method-handler[$method] $what) {
+        $-method-handler[$method][$what] $pkg $cfg
+      } else {
+        fail "No handler for '"$what"' defined in method '"$method"'"
+      }
+    } else {
+      fail "No handler defined for method '"$method"', specified in in config file "(-domain-config-file $dom)
```

Comment:
This message can be made a bit terser to `Unknown method $method`, as the concept of handler is not very relevant here. Also, `in` appeared twice :)